### PR TITLE
[WIP #73] Allow more general block structures

### PR DESCRIPTION
**Pull request by schillic**
See #73.

* [x] new options `:vars`, `:partition`, `:block_types`
* [x] reworked analysis to use the new options
* [x] split `block_types` option for decomposition and overapproximation -> add another argument to the `overapproximate` function (the block index)
* [x] rework projection:
  Add a new option `set_type_proj`, which defaults to `set_type` if it is given. If it is not given, but `block_types` is given, we compute a `set_type_proj` as the "most precise" set type that occurs. (same for epsilon of course)
* [x] adapt unit tests
* [x] ignore other blocks during analysis
* [x] add mapping from input dimension to output dimension
* [x] eliminate unused dimensions in properties
* [x] get rid of single block analysis
* [x] make sure that properties are n-dimensional
* [x] check `block_types`
* [x] testing: compare to master; in particular test all different function versions

To be done after merging:
* synthesize `:partition` and add sanity check for `:partition` (see #92)
* adapt benchmarks

### CODE REVIEW

**Comment by schillic** [comment]
Currently this is experimental and I had to break several things.
You can do this now:
```julia
using Reachability
A = randn(4, 4); X0 = BallInf(ones(4), 0.1);
s = solve(ContinuousSystem(A, X0), :T=>0.1, :set_type=>Interval, :blocks=>[1,2,3,4], :blocks_row=>[1:1, 2:2, 3:3, 4:4], :blocks_col=>[1:1, 2:2, 3:3, 4:4], :apply_projection=>false);
```

File: src/Properties/check_explicit_blocks.jl [hunk 1]
```diff
@@ -23,59 +23,61 @@ OUTPUT:
 The first time index where the property is violated, and 0 if the property is satisfied.
 =#
 
+# helper function
+@inline G0(bi::AbstractVector{Int}) =
```
Comment (mforets) [inline src/Properties/check_explicit_blocks.jl]: `n` as a parameter?

**Comment by schillic** [comment]
Does the sparse (ignore blocks not occurring in `blocks`) analysis make sense for property checking?
If the property involves a linear combination, we would have to use the map to compute the correct dimensions for each check, right?

**Comment by mforets** [comment]
seems to me that `:block_types` is being ignored; tried adding it to `problem_options` inside Building.

**Comment by schillic** [comment]
> seems to me that :block_types is being ignored; tried adding it to problem_options inside Building.

fixed :+1:

**Pull request merged**